protocols/autonat/src/v2/client/behaviour.rs

```diff
@@ -0,0 +1,441 @@
+        _: &Multiaddr,
+        _: &Multiaddr,
+    ) -> Result<<Self as NetworkBehaviour>::ConnectionHandler, ConnectionDenied> {
+        Ok(Either::Right(dial_back::Handler::new()))
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        _: ConnectionId,
+        _: PeerId,
+        _: &Multiaddr,
+        _: Endpoint,
+        _: PortUse,
+    ) -> Result<<Self as NetworkBehaviour>::ConnectionHandler, ConnectionDenied> {
+        Ok(Either::Left(dial_request::Handler::new()))
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm) {
+        match event {
+            FromSwarm::NewExternalAddrCandidate(NewExternalAddrCandidate { addr }) => {
+                self.address_candidates
+                    .entry(addr.clone())
+                    .or_default()
+                    .score += 1;
+            }
+            FromSwarm::ConnectionEstablished(ConnectionEstablished {
+                peer_id,
+                connection_id,
+                endpoint: _,
+                ..
+            }) => {
+                self.peer_info.insert(
+                    connection_id,
+                    ConnectionInfo {
+                        peer_id,
+                        supports_autonat: false,
+                    },
+                );
+            }
+            FromSwarm::ConnectionClosed(ConnectionClosed {
+                peer_id,
+                connection_id,
+                ..
+            }) => {
+                let info = self
+                    .peer_info
+                    .remove(&connection_id)
+                    .expect("inconsistent state");
+
+                if info.supports_autonat {
+                    tracing::debug!(%peer_id, "Disconnected from AutoNAT server");
+                }
+            }
+            _ => {}
+        }
+    }
+
+    fn on_connection_handler_event(
+        &mut self,
+        peer_id: PeerId,
+        connection_id: ConnectionId,
+        event: <Self::ConnectionHandler as ConnectionHandler>::ToBehaviour,
+    ) {
+        let (nonce, outcome) = match event {
+            Either::Right(IncomingNonce { nonce, sender }) => {
+                let Some((_, info)) = self
+                    .address_candidates
+                    .iter_mut()
+                    .find(|(_, info)| info.is_pending_with_nonce(nonce))
+                else {
+                    let _ = sender.send(Err(std::io::Error::new(
+                        std::io::ErrorKind::InvalidData,
+                        format!("Received unexpected nonce: {nonce} from {peer_id}"),
+                    )));
+                    return;
+                };
+
+                info.status = TestStatus::Received(nonce);
+                tracing::debug!(%peer_id, %nonce, "Successful dial-back");
+
+                if let Err(Err(e)) = sender.send(Ok(())) {
+                    tracing::warn!(%peer_id, %nonce, "Failed to send dial-back response ok to client handler: {e}");
+                }
```

Comment:
I think the only error that can happen here is that the channel is disconnected. In that case, `send` returns you what you put into the channel. Here, we always attempt to send `Ok(())`, meaning the pattern `Err(Err())` is unreachable.

I think we can just ignore this error. If we can't send into the channel, it means the handler has de-allocated the task (probably a timeout). Similar to above, I'd just do `let _ = sender.send`.

Reply:
That's a good point. I will change that.